design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
+
+In headless, the session is opened when starting the ksqlDB server with a SQL script. The variables stay alive until the server is
+killed.
+
+Variables are not persisted anywhere. They are living in memory as long as the session is not terminated.
+
+#### RUN SCRIPT command
+
+If a `RUN SCRIPT` command is executed in a CLI session, then any variable defined there will be local to the execution context of the
+script. The variables will override any variable previously defined in the session. When the script is completed, all local variables
+defined will be destroyed. This ensures users do not override their session variables by accident when running `RUN SCRIPT` commands.
+
+#### The SOURCE command
+
+To allow users load variables and keep their scope in the session, then the new `SOURCE` command will be added. This will override
+any previously defined variables, and will be destroyed until the user leaves the session.
+
+Syntax:
+```
+SOURCE '<script_file>';
+```
+
+The `SOURCE` command will work like a `RUN SCRIPT`. If other SQL statements are in the file, then those statements will also be
+executed, thus to avoid limiting users from using DDL/DML commands while defining variables at the same time from one single script.
+
+`SOURCE` will be used only on the CLI and headless mode. REST API will not accept it because files may be out of the scope of the server filesystem.
+
+### 3. Referencing substitution variables
+
+Variables will be referenced by wrapping the variable between `${}` characters (i.e. `${replicas}`).
+
+Noe: We need to make changes in QTT tests to use `${}` references instead of `{}` for consistency.
+
+Example:
+```
+ksql> DEFINE format = 'AVRO';
+ksql> DEFINE replicas = '3';
+ksql> CREATE STREAM stream1 (id INT) WITH (kafka_topic='stream1', value_format='${format}', replicas=${replicas});
+```
+
+Substitution will not attempt to add single-quotes to the values. Users will need to know the data type to use when using a variable. The ksqlDB
+server will be in charge of parsing and failing if the type is not allowed.
+
+Note: Variables are case-insensitive. A reference to `${replicas}` is the same as `${REPLICAS}`.
+
+### 4. Context for substitution variables
+
+Variable substitution will be allowed in specific SQL statements. They can be used to replace text and non-text literals, and identifiers such as
+column names and stream/table names. Variables cannot be used as reserved keywords.
+
+For instance:
```

Comment:
what if I don't want a literal `${foo}` to be replaced? e.g. I actually want `${foo}` to be the value in the statement (`SELECT '${foo}' as dolla_foo FROM bar`)? do i need to turn substitution off and then on after I issue the statement?

Reply:
Yes,
```
ksql> set variable-substitution off;
ksql> SELECT '${foo}' as dolla_foo FROM bar;
ksql> set variable-substitution on;
```

I was playing with escape characters, but the `\${foo}` fails, and the `\\${foo}` results in `SELECT '\${foo}'`. 

I wanted to be fancier and behave like Mysql and Posgtres. They do not substitute variables inside quotes. For instance:
```
ksql> CREATE ... WITH (kafka_topic=@topicName);
or
ksql> CREATE ... WITH (kafka_topic=:topicName);
```
I would have to trick the `StringSubstitutor` a little bit to reject replacing variables inside quotes, it's not a problem. The problem I had was how to replace this:
```
ksql> define colName = 'name';
ksql> SELECT id, ${colName} FROM stream;
```
would become:
```
ksql> SELECT id, "name" FROM stream;   // valid statement 
or
ksql> SELECT id, name FROM stream;     // valid statement
```
Hard to know.  You can do prepared statements in Mysql, but that's very hacky.
https://stackoverflow.com/questions/4428761/mysql-field-name-from-variable

Reply:
I don't think implementation should drive the design. We should define what design we want here in the KLIP, and then if it's tough to implement we can issue a known bug report and fix it later.

cc @colinhicks do you think we need to hammer this issue out before we approve the KLIP?

Reply:
```
ksql> set variable-substitution off;
ksql> SELECT '${foo}' as dolla_foo FROM bar;
ksql> set variable-substitution on;
```

This seems like a reasonable limitation to me.